### ISSUE TITLE
DirectionalLight3D shadow flickers when the camera moves or changes FOV

### ISSUE THREAD

**Issue by rustwork** [godotengine/godot#72015]
### Godot version

4.0 Beta 14

### System information

Windows 10, Vulkan, Nvidia RTX 3060 Ti

### Issue description

If the camera's FOV is animated (for example while sprinting in an FPS game) the shadow flickers. This is most noticeable when the shadow is cast through high frequency alpha tested materials like the leaves of trees.

This problem is made worse as soon as the player gets a bit further from the world origin, likely due to some floating point precision problems. At a relatively small distance of only 500m from the origin there is noticeable flicker of foliage shadows even when the camera is simply moving around normally. It becomes extreme as soon as the camera's FOV changes.

This makes it really difficult to have even a modestly sized outdoor environment without shadow issues.

https://user-images.githubusercontent.com/2095704/214502579-8eb2ea3b-7ca1-457d-8d16-b75c410a567e.mp4

The directional light's settings are pretty standard:

![image](https://user-images.githubusercontent.com/2095704/214502977-2f68ca34-d6c7-4a74-a6f9-4cc1bc97d0ad.png)

### Steps to reproduce

**_From scratch_**
- Create a plane with noisy alpha tested material (like foliage leaves)
- Create a directional light with shadow
- Move parent node 500m from the origin
- Animate the camera's position, rotation and FOV

**_In example project_**
- Run project
- Use WASD keys and mouse to move the camera
- Hold SHIFT key to change FOV

### Minimal reproduction project

[shadow_flicker_bug.zip](https://github.com/godotengine/godot/files/10497038/shadow_flicker_bug.zip)

**Comment by clayjohn** [comment]
Which renderer are you using?

**Comment by Calinou** [comment]
This is an expected issue, which I could reproduce while adding dynamic FOV (depending on car speed) to the Truck Town demo. I worked around this by not changing the camera FOV at low enough speeds. This makes the issue harder to notice as shadow flicker will only occur at high speeds.

**Comment by rustwork** [comment]
@Calinou 
Your maximum FOV proposal sounds like a nice solution for most FOV animation use cases.

To be clear, the shadows in my example are not cast through an Alpha Hash material, it's through an Alpha Scissor material. Changing it to Alpha Hash does indeed make it _very_ jittery. I actually like the quality of the Alpha Scissor shadows, my issue is that even that jitters when the camera is a few hundred meters from the world origin.

Turning off **Blend Splits** makes no difference, it still flickers the same.

**Comment by Calinou** [comment]
> Has there been any progress on this?

To my knowledge, nobody is currently working on this. The comment I wrote above still stands and could be implemented.

**Comment by NuclearPhoenixx** [comment]
> This is an expected issue, which I could reproduce while adding dynamic FOV (depending on car speed) to the [Truck Town demo](https://github.com/godotengine/godot-demo-projects/tree/4.0-dev/3d/truck_town). I worked around this by not changing the camera FOV at low enough speeds. This makes the issue harder to notice as shadow flicker will only occur at high speeds.

I was wondering why this was happening to my shadows, now I know why. Thanks for the comment, I would've never guessed that to be the reason. I second a "maximum camera FOV" workaround.

**Comment by TokisanGames** [comment]
I see, thanks for the tip. The camera is at (1877, 171, 2104), which is not very far at all for an 8k shadow map. 16k maps don't help. It doesn't flicker at (3, 5, 8). It starts flickering around 512-768 and gets progressively worse. It also occurs in 4.5.1 and in the editor. 

We can recenter our 4k world around the origin which will help some. But this seems like a severe limitation. Is the intention for Godot to only support tiny worlds? We support terrains as large as 65.5km per side. We're not going to have a commercially viable engine if shadows are flickering violently beyond 2k.

@clayjohn Is it possible to have the shadows calculated from the camera rather than the origin?

**Comment by clayjohn** [comment]
> @clayjohn Is it possible to have the shadows calculated from the camera rather than the origin?

I'm not sure what you mean. They aren't calculated from the camera or from the origin.

**Comment by TokisanGames** [comment]
@clayjohn Please see the discussion and my video above. Calinou suggested shadow maps are calculated from the origin. And my experience shows severe shadow flickering 2k away from the origin and none near the origin with the same opaque object, in the editor or in game. So whatever is calculated from the origin, can it be moved to the camera instead?

**Comment by Image-unavailable** [comment]
I played around a bit with the MRP 8000 units away from the origin, with low-res one-cascade shadowmaps to see what was going on.

https://github.com/user-attachments/assets/30587a91-7120-46f1-bbe2-d55f17b1addc

My feeling is that whatever code is responsible for snapping positions to shadowmap pixels (to prevent shadow flicker when view moves) is acting up and making the shadowmap camera shake when the main camera moves or rotates.

**Comment by TokisanGames** [comment]
@clayjohn Indeed, two separate issues. If you want me to make a separate ticket, let me know.

Disabling 16-bit shadows doesn't help.

Here's my MRP. This is a copy of my project.godot, the mesh sampled in the video and two scenes:
* flicker.tscn - exact position and settings of the mesh and DirectionalLight3D (Ground1878)
* no_flicker.tscn - new positions and new directional light

It appears related to:
1. `directional_shadow_max_distance`. Mine is set to 400. When set above 100, AND far from 0, 0, 0, the shadows flicker progressively worse. 400 near 0, 0, 0 looks fine. 100 near 1878 still flickers a little.
2. Z rotation on the DirectionalLight3D. 77.7 flickers a bit. 0 is much less. Not an issue near 0, 0, 0.

[test-shadow-flickers.zip](https://github.com/user-attachments/files/24460397/test-shadow-flickers.zip)

**Comment by Image-unavailable** [comment]
I have a full fix for the flickering that occurs when rotating the camera. It fully removes flicker caused by camera rotation at any distance from the origin, so long as the camera doesn't move.

However fully fixing flicker from movement is more complicated, as even doing all computations with `double` doesn't solve it, I suspect that's because once the shadow camera is far enough from the origin, there simply aren't float values that accurately represent the position the shadow camera should snap to in order to keep shadows flicker-free.

**Comment by Image-unavailable** [comment]
...Nevermind, it fixes the flickering but it seems like other stuff breaks instead. :|

**Comment by Image-unavailable** [comment]
PR is up for an improvement to directional shadow stability! #119705
